File: windows/src/CodexAyar.cs
using System.Text.Json.Serialization;

namespace YzlabKurucu;

/// Codex `config.toml`'a YapayZekaLab yonlendirmesini ISARETLI bloklarla ekler / kaldirir.
///
/// NEDEN PROFIL DEGIL: Codex MASAUSTU uygulamasi `-p yzlab` profilini okumaz (2026-09-15 musteri
/// kamer5561: kurucu "Kuruldu" dedi ama masaustu kendi ChatGPT hesabiyla gitti → "hakkiniz yok").
/// Sahibin CALISAN masaustu kurulumu ana config.toml'da model_provider + [model_providers.yapayzekalab]
/// (experimental_bearer_token) kullaniyor; auth.json'daki ChatGPT girisi duruyor.
///
/// Saf metin islemi. macOS karsiligi `CodexAyar.swift` BIREBIR ayni kurallar.
public static class CodexAyar
{
    public const string UstBas = "# >>> YapayZekaLab kurucu: ust ayarlar (Geri Al kaldirir; elle degistirme)";
    public const string UstSon = "# <<< YapayZekaLab kurucu: ust ayarlar";
    public const string SagBas = "# >>> YapayZekaLab kurucu: saglayici";
    public const string SagSon = "# <<< YapayZekaLab kurucu: saglayici";
    public const string Saglayici = "yapayzekalab";

    /// Kurulumdan ONCE dosyada olup bizim kaldirdigimiz seyler. Geri Al bunlari geri koyar.
    public sealed class Yedek
    {
        [JsonPropertyName("dosyaYoktu")] public bool DosyaYoktu { get; set; }
        [JsonPropertyName("ustSatirlar")] public List<string> UstSatirlar { get; set; } = new();
        [JsonPropertyName("saglayiciBlogu")] public List<string> SaglayiciBlogu { get; set; } = new();
        [JsonPropertyName("yonetilenAnahtarlar")] public List<string> YonetilenAnahtarlar { get; set; } = new();
    }

    public sealed record Sablon(List<string> UstSatirlar, List<string> Anahtarlar, List<string> SaglayiciSatirlar);

    // ── Satir yardimcilari ────────────────────────────────────────────────

    public static (List<string> Satirlar, string Nl) SatirlaraAyir(string? metin)
    {
        var s = metin ?? "";
        if (s.Length > 0 && s[0] == '﻿') s = s[1..];
        var nl = s.Contains("\r\n") ? "\r\n" : "\n";
        var parca = s.Replace("\r\n", "\n").Split('\n').ToList();
        if (parca.Count > 0 && parca[^1] == "") parca.RemoveAt(parca.Count - 1);
        return (parca, nl);
    }

    public static bool BosMu(string l) => string.IsNullOrWhiteSpace(l);
    public static bool BasliMi(string l) => l.TrimStart().StartsWith('[');

    public static string? Anahtar(string l)
    {
        var t = l.Trim();
        if (t.Length == 0 || t[0] == '#' || t[0] == '[') return null;
        var e = t.IndexOf('=');
        if (e <= 0) return null;
        var k = t[..e].Trim();
        if (k.Length >= 2 && ((k[0] == '"' && k[^1] == '"') || (k[0] == '\'' && k[^1] == '\'')))
            k = k[1..^1];
        return k.Length == 0 ? null : k;
    }

    public static string? BaslikAdi(string l)
    {
        var t = l.Trim();
        if (!t.StartsWith('[')) return null;
        var kapa = t.LastIndexOf(']');
        if (kapa < 0) return null;
        var ic = t[..(kapa + 1)].Trim('[', ']');
        return ic.Replace(" ", "").Replace("\"", "").Replace("'", "");
    }

    public static bool SaglayiciBasligiMi(string l)
    {
        var ad = BaslikAdi(l);
        if (ad is null) return false;
        return ad == "model_providers." + Saglayici || ad.StartsWith("model_providers." + Saglayici + ".");
    }

    public static List<string> IsaretlileriSil(List<string> L)
    {
        var outL = new List<string>();
        var i = 0;
        while (i < L.Count)
        {
            var t = L[i].Trim();
            if (t == UstBas || t == SagBas)
            {
                var son = t == UstBas ? UstSon : SagSon;
                var j = -1;
                for (var x = i + 1; x < L.Count; x++) if (L[x].Trim() == son) { j = x; break; }
                i = j >= 0 ? j + 1 : i + 1;
                continue;
            }
            if (t == UstSon || t == SagSon) { i++; continue; }
            outL.Add(L[i]);
            i++;
        }
        return outL;
    }

    public static (List<string> Ust, List<string> Kalan, List<string> YUst, List<string> YSag)
        Ayikla(List<string> L, HashSet<string> anahtarlar)
    {
        var ilk = L.FindIndex(BasliMi);
        if (ilk < 0) ilk = L.Count;
        var ust = new List<string>(); var yUst = new List<string>();
        for (var i = 0; i < ilk; i++)
        {
            var k = Anahtar(L[i]);
            if (k is not null && anahtarlar.Contains(k)) yUst.Add(L[i]); else ust.Add(L[i]);
        }
        var kalan = new List<string>(); var ySag = new List<string>();
        var siliyor = false;
        for (var i = ilk; i < L.Count; i++)
        {
            if (BasliMi(L[i])) siliyor = SaglayiciBasligiMi(L[i]);
            if (siliyor) ySag.Add(L[i]); else kalan.Add(L[i]);
        }
        while (ySag.Count > 0 && BosMu(ySag[^1])) ySag.RemoveAt(ySag.Count - 1);
        while (ust.Count > 0 && BosMu(ust[0])) ust.RemoveAt(0);
        return (ust, kalan, yUst, ySag);
    }

    // ── Sablon ────────────────────────────────────────────────────────────

    /// Doldurulmus profil sablonunu ust anahtarlara + saglayici tablosuna ayirir.
    /// [tools] gibi DIGER tablolar ALINMAZ: musterinin kendi tablosuyla cakisip TOML'u bozar.
    public static Sablon SablonuAyir(string dolu)
    {
        var L = SatirlaraAyir(dolu).Satirlar;
        var ust = new List<string>(); var anah = new List<string>(); var sag = new List<string>();
        bool baslikGoruldu = false, saglayicida = false;
        foreach (var l in L)
        {
            if (BasliMi(l))
            {
                baslikGoruldu = true;
                saglayicida = SaglayiciBasligiMi(l);
                if (saglayicida) sag.Add(l.Trim());
                continue;
            }
            var t = l.Trim();
            if (!baslikGoruldu)
            {
                var k = Anahtar(l);
                if (k is not null) { ust.Add(t); anah.Add(k); }
            }
            else if (saglayicida && t.Length > 0 && t[0] != '#')
            {
                sag.Add(t);
            }
        }
        return new Sablon(ust, anah, sag);
    }

    // ── Uygula / Geri al / Dogrula ────────────────────────────────────────

    public static (string Metin, Yedek Yedek) Uygula(string? mevcut, Sablon sablon)
    {
        var (ham, nl) = SatirlaraAyir(mevcut);
        var a = Ayikla(IsaretlileriSil(ham), new HashSet<string>(sablon.Anahtarlar));
        var outL = new List<string> { UstBas };
        outL.AddRange(sablon.UstSatirlar);
        outL.Add(UstSon);
        var govde = a.Ust.Concat(a.Kalan).ToList();
        if (govde.Count > 0) { outL.Add(""); outL.AddRange(govde); }
        while (outL.Count > 0 && BosMu(outL[^1])) outL.RemoveAt(outL.Count - 1);
        outL.Add(""); outL.Add(SagBas); outL.AddRange(sablon.SaglayiciSatirlar); outL.Add(SagSon);
        var yedek = new Yedek
        {
            DosyaYoktu = mevcut is null,
            UstSatirlar = a.YUst,
            SaglayiciBlogu = a.YSag,
            YonetilenAnahtarlar = sablon.Anahtarlar.ToList(),
        };
        return (string.Join(nl, outL) + nl, yedek);
    }

    /// null → dosya kurulumdan once YOKTU ve geriye bir sey kalmadi (sil).
    public static string? GeriAl(string? mevcut, Yedek yedek)
    {
        var (ham, nl) = SatirlaraAyir(mevcut);
        var a = Ayikla(IsaretlileriSil(ham), new HashSet<string>(yedek.YonetilenAnahtarlar));
        var govde = a.Ust.Concat(a.Kalan).ToList();
        while (govde.Count > 0 && BosMu(govde[0])) govde.RemoveAt(0);
        var outL = new List<string>(yedek.UstSatirlar);
        if (outL.Count > 0 && govde.Count > 0 && BasliMi(govde[0])) outL.Add("");
        outL.AddRange(govde);
        while (outL.Count > 0 && BosMu(outL[^1])) outL.RemoveAt(outL.Count - 1);
        if (yedek.SaglayiciBlogu.Count > 0)
        {
            if (outL.Count > 0) outL.Add("");
            outL.AddRange(yedek.SaglayiciBlogu);
        }
        if (outL.All(BosMu)) return yedek.DosyaYoktu ? null : "";
        return string.Join(nl, outL) + nl;
    }

    /// Codex'i ACILMAZ yapacak hatalari yakalar. null = sorun yok.
    public static string? Dogrula(string metin)
    {
        var L = SatirlaraAyir(metin).Satirlar;
        var ilk = L.FindIndex(BasliMi);
        if (ilk < 0) ilk = L.Count;
        var gorulen = new HashSet<string>();
        for (var i = 0; i < ilk; i++)
        {
            var k = Anahtar(L[i]);
            if (k is null) continue;
            if (!gorulen.Add(k)) return "config.toml'da ayni ust anahtar iki kez: " + k;
        }
        var basliklar = new HashSet<string>();
        for (var i = ilk; i < L.Count; i++)
        {
            if (!BasliMi(L[i]) || L[i].TrimStart().StartsWith("[[")) continue;
            var ad = BaslikAdi(L[i]);
            if (ad is null) continue;
            if (!basliklar.Add(ad)) return "config.toml'da ayni tablo iki kez: [" + ad + "]";
        }
        if (!basliklar.Contains("model_providers." + Saglayici)) return "saglayici tablosu yazilamadi";
        string? saglayiciSatiri = null;
        for (var i = 0; i < ilk; i++) if (Anahtar(L[i]) == "model_provider") { saglayiciSatiri = L[i]; break; }
        if (saglayiciSatiri is null || !saglayiciSatiri.Contains("\"" + Saglayici + "\""))
            return "model_provider yapayzekalab degil";
        return null;
    }
}

File: macos/Sources/YzlabKurucu/CodexAyar.swift
import Foundation

/// Codex `config.toml`'a YapayZekaLab yönlendirmesini İŞARETLİ bloklarla ekler / kaldırır.
///
/// NEDEN PROFİL DEĞİL: Codex MASAÜSTÜ uygulaması `-p yzlab` profilini okumaz. 2026-09-15 müşteri
/// kamer5561: kurucu "Kuruldu" dedi (testi profil üzerinden bize geldi) ama masaüstü kendi ChatGPT
/// hesabıyla gitti → "hakkınız yok". Sahibin ÇALIŞAN masaüstü kurulumu da ana config.toml'da
/// `model_provider = "yapayzekalab"` + `[model_providers.yapayzekalab]` (experimental_bearer_token)
/// kullanıyor; auth.json'daki ChatGPT girişi duruyor. Bu modül tam olarak onu yazar.
///
/// Saf metin işlemi (dosya G/Ç yok). Windows karşılığı `CodexAyar.cs` BİREBİR aynı kurallar.
enum CodexAyar {
    static let ustBas = "# >>> YapayZekaLab kurucu: ust ayarlar (Geri Al kaldirir; elle degistirme)"
    static let ustSon = "# <<< YapayZekaLab kurucu: ust ayarlar"
    static let sagBas = "# >>> YapayZekaLab kurucu: saglayici"
    static let sagSon = "# <<< YapayZekaLab kurucu: saglayici"
    static let saglayici = "yapayzekalab"

    /// Kurulumdan ÖNCE dosyada olup bizim kaldırdığımız şeyler. Geri Al bunları geri koyar.
    struct Yedek: Codable, Equatable {
        var dosyaYoktu: Bool
        var ustSatirlar: [String]
        var saglayiciBlogu: [String]
        var yonetilenAnahtarlar: [String]
    }

    /// Manifestteki (doldurulmuş) şablondan çıkarılan parçalar.
    struct Sablon: Equatable {
        var ustSatirlar: [String]
        var anahtarlar: [String]
        var saglayiciSatirlar: [String]
    }

    // MARK: - Satır yardımcıları

    static func satirlaraAyir(_ metin: String) -> (satirlar: [String], nl: String) {
        var s = metin
        if s.hasPrefix("\u{FEFF}") { s.removeFirst() }
        let nl = s.contains("\r\n") ? "\r\n" : "\n"
        var parca = s.replacingOccurrences(of: "\r\n", with: "\n").components(separatedBy: "\n")
        if parca.last == "" { parca.removeLast() }
        return (parca, nl)
    }

    static func bosMu(_ l: String) -> Bool { l.trimmingCharacters(in: .whitespaces).isEmpty }
    static func basliMi(_ l: String) -> Bool { l.trimmingCharacters(in: .whitespaces).hasPrefix("[") }

    /// `anahtar = deger` satırının anahtarı (yorum/başlık/boş satırda nil).
    static func anahtar(_ l: String) -> String? {
        let t = l.trimmingCharacters(in: .whitespaces)
        guard !t.isEmpty, !t.hasPrefix("#"), !t.hasPrefix("["), let e = t.firstIndex(of: "=") else { return nil }
        var k = t[..<e].trimmingCharacters(in: .whitespaces)
        if k.count >= 2, (k.hasPrefix("\"") && k.hasSuffix("\"")) || (k.hasPrefix("'") && k.hasSuffix("'")) {
            k = String(k.dropFirst().dropLast())
        }
        return k.isEmpty ? nil : k
    }

    /// `[ model_providers . "yapayzekalab" ]` → `model_providers.yapayzekalab` (boşluk/tırnak atılır).
    static func baslikAdi(_ l: String) -> String? {
        let t = l.trimmingCharacters(in: .whitespaces)
        guard t.hasPrefix("["), let kapa = t.lastIndex(of: "]") else { return nil }
        let ic = String(t[t.startIndex...kapa]).trimmingCharacters(in: CharacterSet(charactersIn: "[]"))
        return ic.replacingOccurrences(of: " ", with: "")
            .replacingOccurrences(of: "\"", with: "")
            .replacingOccurrences(of: "'", with: "")
    }

    static func saglayiciBasligiMi(_ l: String) -> Bool {
        guard let ad = baslikAdi(l) else { return false }
        return ad == "model_providers.\(saglayici)" || ad.hasPrefix("model_providers.\(saglayici).")
    }

    /// Bizim işaretli bloklarımızı (başlangıç..bitiş dahil) atar. Sonu olmayan işaret → yalnız o satır.
    static func isaretlileriSil(_ L: [String]) -> [String] {
        var out: [String] = []
        var i = 0
        while i < L.count {
            let t = L[i].trimmingCharacters(in: .whitespaces)
            if t == ustBas || t == sagBas {
                let son = (t == ustBas) ? ustSon : sagSon
                if let j = L[(i + 1)...].firstIndex(where: { $0.trimmingCharacters(in: .whitespaces) == son }) {
                    i = j + 1
                    continue
                }
                i += 1
                continue
            }
            if t == ustSon || t == sagSon { i += 1; continue }
            out.append(L[i])
            i += 1
        }
        return out
    }

    /// Yönetilen üst anahtarları ve yapayzekalab sağlayıcı tablolarını ayıklar (işaret DIŞINDA olsalar da).
    static func ayikla(_ L: [String], anahtarlar: Set<String>)
        -> (ust: [String], kalan: [String], yUst: [String], ySag: [String]) {
        let ilk = L.firstIndex(where: basliMi) ?? L.count
        var ust: [String] = [], yUst: [String] = []
        for l in L[..<ilk] {
            if let k = anahtar(l), anahtarlar.contains(k) { yUst.append(l) } else { ust.append(l) }
        }
        var kalan: [String] = [], ySag: [String] = []
        var siliyor = false
        for l in L[ilk...] {
            if basliMi(l) { siliyor = saglayiciBasligiMi(l) }
            if siliyor { ySag.append(l) } else { kalan.append(l) }
        }
        while let s = ySag.last, bosMu(s) { ySag.removeLast() }
        while let f = ust.first, bosMu(f) { ust.removeFirst() }
        return (ust, kalan, yUst, ySag)
    }

    // MARK: - Şablon

    /// Doldurulmuş profil şablonunu üst anahtarlara + sağlayıcı tablosuna ayırır.
    /// `[tools]` gibi DİĞER tablolar ALINMAZ: müşterinin kendi tablosuyla çakışıp TOML'u bozar.
    static func sablonuAyir(_ dolu: String) -> Sablon {
        let L = satirlaraAyir(dolu).satirlar
        var ust: [String] = [], anah: [String] = [], sag: [String] = []
        var baslikGoruldu = false, saglayicida = false
        for l in L {
            if basliMi(l) {
                baslikGoruldu = true
                saglayicida = saglayiciBasligiMi(l)
                if saglayicida { sag.append(l.trimmingCharacters(in: .whitespaces)) }
                continue
            }
            let t = l.trimmingCharacters(in: .whitespaces)
            if !baslikGoruldu {
                if let k = anahtar(l) { ust.append(t); anah.append(k) }
            } else if saglayicida, !t.isEmpty, !t.hasPrefix("#") {
                sag.append(t)
            }
        }
        return Sablon(ustSatirlar: ust, anahtarlar: anah, saglayiciSatirlar: sag)
    }

    // MARK: - Uygula / Geri al / Doğrula

    /// `mevcut` nil → dosya yoktu. Dönen `yedek` bu çağrıda kaldırılanlardır; çağıran İLK kurulumun
    /// yedeğini korumalıdır (yeniden kurmak yedeği ezmesin).
    static func uygula(_ mevcut: String?, _ sablon: Sablon) -> (metin: String, yedek: Yedek) {
        let (ham, nl) = satirlaraAyir(mevcut ?? "")
        let a = ayikla(isaretlileriSil(ham), anahtarlar: Set(sablon.anahtarlar))
        var out = [ustBas] + sablon.ustSatirlar + [ustSon]
        let govde = a.ust + a.kalan
        if !govde.isEmpty { out.append(""); out += govde }
        while let s = out.last, bosMu(s) { out.removeLast() }
        out += ["", sagBas] + sablon.saglayiciSatirlar + [sagSon]
        let yedek = Yedek(dosyaYoktu: mevcut == nil, ustSatirlar: a.yUst,
                          saglayiciBlogu: a.ySag, yonetilenAnahtarlar: sablon.anahtarlar)
        return (out.joined(separator: nl) + nl, yedek)
    }

    /// Kurulum öncesine döner. nil → dosya kurulumdan önce YOKTU ve geriye bir şey kalmadı (sil).
    /// Kurulumdan SONRA eklenenler (ör. Codex'in yazdığı `[projects.*]`) korunur.
    static func geriAl(_ mevcut: String?, _ yedek: Yedek) -> String? {
        let (ham, nl) = satirlaraAyir(mevcut ?? "")
        let a = ayikla(isaretlileriSil(ham), anahtarlar: Set(yedek.yonetilenAnahtarlar))
        var govde = a.ust + a.kalan
        while let f = govde.first, bosMu(f) { govde.removeFirst() }
        var out = yedek.ustSatirlar
        if !out.isEmpty, let f = govde.first, basliMi(f) { out.append("") }
        out += govde
        while let s = out.last, bosMu(s) { out.removeLast() }
        if !yedek.saglayiciBlogu.isEmpty {
            if !out.isEmpty { out.append("") }
            out += yedek.saglayiciBlogu
        }
        if out.allSatisfy(bosMu) { return yedek.dosyaYoktu ? nil : "" }
        return out.joined(separator: nl) + nl
    }

    /// Codex'i AÇILMAZ yapacak hataları yakalar (TOML'da tekrar eden anahtar/tablo). nil = sorun yok.
    static func dogrula(_ metin: String) -> String? {
        let L = satirlaraAyir(metin).satirlar
        let ilk = L.firstIndex(where: basliMi) ?? L.count
        var gorulen = Set<String>()
        for l in L[..<ilk] {
            guard let k = anahtar(l) else { continue }
            if gorulen.contains(k) { return "config.toml'da ayni ust anahtar iki kez: \(k)" }
            gorulen.insert(k)
        }
        var basliklar = Set<String>()
        for l in L[ilk...] where basliMi(l) {
            if l.trimmingCharacters(in: .whitespaces).hasPrefix("[[") { continue }
            guard let ad = baslikAdi(l) else { continue }
            if basliklar.contains(ad) { return "config.toml'da ayni tablo iki kez: [\(ad)]" }
            basliklar.insert(ad)
        }
        guard basliklar.contains("model_providers.\(saglayici)") else { return "saglayici tablosu yazilamadi" }
        let saglayiciSatiri = L[..<ilk].first { anahtar($0) == "model_provider" }
        guard let s = saglayiciSatiri, s.contains("\"\(saglayici)\"") else { return "model_provider yapayzekalab degil" }
        return nil
    }
}
